networkx/utils/decorators.py

```diff
@@ -392,84 +355,868 @@ def random_array(dims, random_state=1):
+    use the `@argmap(...)` idiom to save space here.
+
+    As an example, the decorated function::
+
+        @argmap(sum, "x", "z")
+        def foo(x, y, z):
+            return x - y + z
+
+    is syntactic sugar for::
 
-        # Create a numpy.random.RandomState instance
-        random_state = create_py_random_state(random_state_arg)
+        def foo(x, y, z):
+            x = sum(x)
+            z = sum(z)
+            return x - y + z
 
-        # args is a tuple, so we must convert to list before modifying it.
-        new_args = list(args)
-        new_args[random_state_index] = random_state
-        return func(*new_args, **kwargs)
+    and is equivalent to (using argument indexes)::
 
-    return _random_state
+        @argmap(sum, "x", 2)
+        def foo(x, y, z):
+            return x - y + z
+
+    or::
+
+        @argmap(sum, "z", 0)
+        def foo(x, y, z):
+            return x - y + z
+
+    Transforming functions can be applied to multiple arguments, such as::
+
+        def swap(x, y):
+            return y, x
+
+        # the 2-tuple tells argmap that the map `swap` has 2 inputs/outputs.
+        @argmap(swap, ("a", "b")):
+        def foo(a, b, c):
+            return a / b * c
+
+    is equivalent to::
+
+        def foo(a, b, c):
+            a, b = swap(a, b)
+            return a / b * c
+
+    More generally, the applied arguments can be nested tuples of strings or ints.
+    The syntax `@argmap(some_func, ("a", ("b", "c")))` would expect `some_func` to
+    accept 2 inputs with the second expected to be a 2-tuple. It should then return
+    2 outputs with the second a 2-tuple. The returns values would replace input "a"
+    "b" and "c" respectively. Similarly for `@argmap(some_func, (0, ("b", 2)))`.
+
+    Also, note that an index larger than the number of named parameters is allowed
+    so long as a VAR_POSITIONAL input appears, e.g. `*args`. In that case the index
+    extends past the named parameters into the `args` tuple. For example::
+
+        def double(a):
+            return 2 * a
+
+        @argmap(double, 3)
+        def overflow(a, *args):
+            return a, args
+
+        print(overflow(1, 2, 3, 4, 5, 6))  # output is 1, (2, 3, 8, 5, 6)
+
+    **Try Finally**
+
+    Additionally, this `argmap` class can be used to create a decorator that
+    initiates a try...finally block. The decorator must be written to return
+    both the transformed argument and a closing function.
+    This feature was included to enable the `open_file` decorator which might
+    need to close the file or not depending on whether it had to open that file.
+    This feature uses the keyword-only `try_finally` argument to `@argmap`.
+
+    For example this map opens a file and then makes sure it is closed::
+
+        def open_file(fn):
+            f = open(fn)
+            return f, lambda: f.close()
+
+    The decorator applies that to the function `foo`::
+
+        @argmap(open_file, "file", try_finally=True)
```

Comment:
I'm curious about this choice of API.  I find it confusing that the `open_file` function does not transform arguments, as it does elsewhere with `@argmap`.  I'm wondering about the feasibility of the following:

```
def open_file(fn):
    return open(fn)

@argmap(open_file, "file", try_finally=lambda f: f.close())
```

Also, I'm curious whether this can be handled with something like:

```
def open_file(fn):
    with open(fn) as f:
        yield f
    # here, any other finalization operations needed
```

Then `@argmap(open_file, "file")`.

Reply:
Previously, I used an alternative constructor for the class.  Dan didn't like that so I moved it to a keyword argument.  Your proposal here doesn't capture whether or not the file needs to be closed, and may close a filehandle that the user wants to keep open.

Reply:
That's right.  My later suggestion was to simply use a context manager that does the opening, or that creates a temporary file (or whatever is needed).

Reply:
The point of the `try_finally` logic is to close the file when it is appropriate. Opening the file is not the issue.

Also, see #2298 which discusses removing `open_file`.  
But that’s a separate issue from this PR I think…  The discussion there left off with: open_file is too complicated and confusing, but we still want to keep strings (and pathlib.Path objects) as valid ways to specify what file to use…
The tricky part is making sure we close files that we create and leave open files the user creates.

Reply:
Still, this should be doable with a custom context manager.  You can, e.g., return a NetworkXFile object that has a .close method that gets called when exiting the context.  The close method can be customized to do whatever (leave the file alone, close files open by nx, etc.).  And agreed that the issue is not relevant to this specific PR.

Reply:
Please move this conversation/duplicate your suggestions to #2298 as it will be easier to follow & preserve there.

Reply:
Oh, but I should address one bit I missed:

> I'm curious about this choice of API. I find it confusing that the open_file function does not transform arguments

It _does_ transform arguments.  It takes a filename argument, and transforms it to an open file (or does nothing to filehandles).